### PR TITLE
SCANGRADLE-143 SCANGRADLE-144 The scanner for Gradle collects files Outside of conventional sonar.sources

### PR DESCRIPTION
**Pull request by ADarko22**

### CODE REVIEW

File: src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java [hunk 1]
```diff
@@ -143,12 +163,74 @@ private void computeSonarProperties(Project project, Map<String, Object> propert
...
+    var scanAllEnabled = "true".equalsIgnoreCase(scanAllValue.trim());
+
+    if (scanAllEnabled) {
+      LOGGER.info("Parameter sonar.gradle.scanAll is enabled. The scanner will attempt to collect additional sources.");
+
+      // Collecting the properties configured in the Gradle build configuration
+      var sonarProps = new SonarProperties(new HashMap<>());
+      actionBroadcastMap.get(project.getPath()).execute(sonarProps);
+
+      var sourcesOrTestsAlreadySet = Stream.of(System.getProperties().keySet(), Utils.loadEnvironmentProperties(System.getenv()).keySet(), sonarProps.getProperties().keySet())
```
Comment (dorian-burihabwa-sonarsource) [inline src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java:185]: Let's have the explicit boolean type

File: src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java [hunk 2]
```diff
@@ -143,12 +163,74 @@ private void computeSonarProperties(Project project, Map<String, Object> propert
...
+        .flatMap(Collection::stream)
+        .anyMatch(k -> ScanProperties.PROJECT_SOURCE_DIRS.equals(k) || ScanProperties.PROJECT_TEST_DIRS.equals(k));
```
Comment (dorian-burihabwa-sonarsource) [inline src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java:187]: let's check if the property ends with rather than equals to have consistency with the [maven implementation](https://github.com/SonarSource/sonar-scanner-maven/blob/5ca170970f60a41068afbae3cf631db0f349a15c/src/main/java/org/sonarsource/scanner/maven/bootstrap/ScannerBootstrapper.java#L169)

File: src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java [hunk 3]
```diff
@@ -143,12 +163,74 @@ private void computeSonarProperties(Project project, Map<String, Object> propert
...
+      .map(Project::getProjectDir)
+      .map(File::toPath)
+      .collect(Collectors.toSet());
+
+    SourceCollector visitor = new SourceCollector(existingSources, skippedDirs, Set.of(), false);
+
+    try {
+      Files.walkFileTree(project.getProjectDir().toPath(), visitor);
+    } catch (IOException e) {
+      LOGGER.warn(String.valueOf(e));
```
Comment (dorian-burihabwa-sonarsource) [inline src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java:218]: Question: Is an early return safer here?

File: src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java [hunk 4]
```diff
@@ -382,7 +464,8 @@ private static class InternalKotlinSourceSet {
     private String name;
     private Collection<File> srcDirs;
 
-    private InternalKotlinSourceSet() {}
+    private InternalKotlinSourceSet() {
+    }
```
Comment (dorian-burihabwa-sonarsource) [inline src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java:468]:
```suggestion
    private InternalKotlinSourceSet() {}
```

File: src/test/groovy/org/sonarqube/gradle/FunctionalTests.groovy [hunk 5]
```diff
@@ -280,12 +280,12 @@ class FunctionalTests extends Specification {
...
-          .forwardOutput()
-          .withArguments('sonarqube', '-Dsonar.scanner.dumpToFile=' + outFile.toAbsolutePath())
-          .withPluginClasspath()
-          .build()
+                .withGradleVersion(gradleVersion)
+                .withProjectDir(testProjectDir.toFile())
+                .forwardOutput()
+                .withArguments('sonarqube', '-Dsonar.scanner.dumpToFile=' + outFile.toAbsolutePath())
+                .withPluginClasspath()
+                .build()
```
Comment (dorian-burihabwa-sonarsource) [inline src/test/groovy/org/sonarqube/gradle/FunctionalTests.groovy:288]: Let's revert these formatting changes

File: src/test/groovy/org/sonarqube/gradle/FunctionalTests.groovy [hunk 6]
```diff
@@ -344,15 +344,138 @@ class FunctionalTests extends Specification {
...
+
+        then:
+        result.task(":sonar").outcome == SUCCESS
+
+        def props = new Properties()
+        props.load(outFile.newDataInputStream())
+        props."sonar.gradle.scanAll" == "true"
+        result.output.contains("Parameter sonar.gradle.scanAll is enabled. The scanner will attempt to collect additional sources.")
+        result.output.contains("Parameter sonar.gradle.scanAll is enabled but the scanner will not collect additional sources because sonar.sources or sonar.tests are already set.")
```
Comment (dorian-burihabwa-sonarsource) [inline src/test/groovy/org/sonarqube/gradle/FunctionalTests.groovy:426]: let's change the end of the message to be more consistent with the Maven implementation by mentioning overriding
```suggestion
        result.output.contains("Parameter sonar.gradle.scanAll is enabled but the scanner will not collect additional sources because sonar.sources or sonar.tests has been overridden..")
```
```suggestion
        result.output.contains("Parameter sonar.gradle.scanAll is enabled but the scanner will not collect additional sources because sonar.sources or sonar.tests are already set.")
```

File: src/test/groovy/org/sonarqube/gradle/FunctionalTests.groovy [hunk 7]
```diff
@@ -344,15 +344,138 @@ class FunctionalTests extends Specification {
...
+
+        where:
+        sonarSourcesOverride | sonarTestsOverride
+        "src"                | null
+        null                 | "test"
+        "src"                | "test"
+    }
+
+    def "scan all is enabled but not applied because of overridden properties in Gradle"() {
```
Comment (dorian-burihabwa-sonarsource) [inline src/test/groovy/org/sonarqube/gradle/FunctionalTests.groovy:435]:
```suggestion
    def "scan all is enabled but not applied because of overridden properties in build configuration"() {
```

File: src/test/groovy/org/sonarqube/gradle/FunctionalTests.groovy [hunk 8]
```diff
@@ -344,15 +344,138 @@ class FunctionalTests extends Specification {
...
+        result.output.contains("Parameter sonar.gradle.scanAll is enabled. The scanner will attempt to collect additional sources.")
+
+        var sources = ((String) props."sonar.sources").split(",")
+        sources.size() == 3
+        sources[0].endsWith("""$testProjectDir/build.gradle""")
+        sources[1].endsWith("""$testProjectDir/gradle.properties""")
+        sources[2].endsWith("""$testProjectDir/settings.gradle""")
+    }
+
+    def "scan all is enabled but not applied"() {
```
Comment (dorian-burihabwa-sonarsource) [inline src/test/groovy/org/sonarqube/gradle/FunctionalTests.groovy:395]:
```suggestion
    def "scan all is enabled but not applied because of overridden properties on the command line"() {
```

File: src/test/groovy/org/sonarqube/gradle/SonarQubePluginTest.groovy [hunk 9]
```diff
@@ -52,910 +52,967 @@ import static spock.util.matcher.HamcrestSupport.expect
...
+        def srcMainJavaPath = normalizePathString((dir as String) + "/src/main/java")
+        def buildFilePath = normalizePathString((dir as String) + "/build.gradle.kts")
+
+        then:
+        assert sources.size() == 2
+        assert normalizePathArray(sources).contains(buildFilePath)
+        assert normalizePathArray(sources).contains(srcMainJavaPath)
+    }
+
+    def "scan all detects scripts only within non skipped submodules"() {
```
Comment (dorian-burihabwa-sonarsource) [inline src/test/groovy/org/sonarqube/gradle/SonarQubePluginTest.groovy:956]: All the lines above should be back to their default formatting

File: src/test/groovy/org/sonarqube/gradle/SonarQubePluginTest.groovy [hunk 10]
```diff
@@ -52,910 +52,967 @@ import static spock.util.matcher.HamcrestSupport.expect
...
+        assert properties[":module2.sonar.sources"] == null
+        assert properties[":module2.:module2:submodule.sonar.sources"] == null
+    }
+
+    private static void setSourceSets(Project project, List<String> mainDirs) {
+        JavaPluginExtension javaExtension = project.getExtensions().getByType(JavaPluginExtension.class)
+        SourceSetContainer sourceSets = javaExtension.getSourceSets();
+        SourceSet mainSourceSet = sourceSets.getByName("main");
+        for (String mainDir : mainDirs)
+            mainSourceSet.getJava().srcDir(mainDir);
```
Comment (dorian-burihabwa-sonarsource) [inline src/test/groovy/org/sonarqube/gradle/SonarQubePluginTest.groovy:1016]:
```suggestion
        for (String mainDir : mainDirs) {
            mainSourceSet.getJava().srcDir(mainDir);
        }
```

**Review by dorian-burihabwa-sonarsource — CHANGES_REQUESTED** [review]

File: src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java (hunk 3 above)
Comment (ADarko22) [reply src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java:225]: We have this behavior on the scanner for Maven as well. We should change it on both projects to fail

File: integrationTests/src/test/java/org/sonarqube/gradle/GradleTest.java [hunk 11]
```diff
@@ -269,20 +270,19 @@ public void testKotlinMultiplatformProject() throws Exception {
...
-    );
+      baseDir.resolve("src/jvmMain/java").toString());
   }
 
   @Test
   public void testKotlinMultiplatformWithSubmoduleProject() throws Exception {
     Assume.assumeTrue("Tests only applies to version 6.8.3 or greater", getGradleVersion().isGreaterThanOrEqualTo("6.8.3"));
-    Properties props = runGradlewSonarSimulationModeWithEnv("/kotlin-multiplatform-with-submodule", emptyMap(), "compileCommonMainKotlinMetadata", "compileKotlinJvm", "compileKotlinMetadata", "compileTestKotlinJvm");
+    Properties props = runGradlewSonarSimulationModeWithEnv("/kotlin-multiplatform-with-submodule", emptyMap(), "compileCommonMainKotlinMetadata", "compileKotlinJvm",
+      "compileKotlinMetadata", "compileTestKotlinJvm");
```
Comment (dorian-burihabwa-sonarsource) [inline integrationTests/src/test/java/org/sonarqube/gradle/GradleTest.java:285]: Let's revert this change
```suggestion
    Properties props = runGradlewSonarSimulationModeWithEnv("/kotlin-multiplatform-with-submodule", emptyMap(), "compileCommonMainKotlinMetadata", "compileKotlinJvm", "compileKotlinMetadata", "compileTestKotlinJvm");
```

File: integrationTests/src/test/java/org/sonarqube/gradle/GradleTest.java [hunk 12]
```diff
@@ -258,7 +258,8 @@ public void testProjectWithConfigurationCacheDoubleExecutionsShouldWork() throws
   @Test
   public void testKotlinMultiplatformProject() throws Exception {
     Assume.assumeTrue("Tests only applies to version 6.8.3 or greater", getGradleVersion().isGreaterThanOrEqualTo("6.8.3"));
-    Properties props = runGradlewSonarSimulationModeWithEnv("/kotlin-multiplatform", emptyMap(), "compileCommonMainKotlinMetadata", "compileKotlinJvm", "compileKotlinMetadata", "compileTestKotlinJvm");
+    Properties props = runGradlewSonarSimulationModeWithEnv("/kotlin-multiplatform", emptyMap(), "compileCommonMainKotlinMetadata", "compileKotlinJvm", "compileKotlinMetadata",
+      "compileTestKotlinJvm");
```
Comment (dorian-burihabwa-sonarsource) [inline integrationTests/src/test/java/org/sonarqube/gradle/GradleTest.java:262]: Let's revert this change
```suggestion
    Properties props = runGradlewSonarSimulationModeWithEnv("/kotlin-multiplatform", emptyMap(), "compileCommonMainKotlinMetadata", "compileKotlinJvm", "compileKotlinMetadata", "compileTestKotlinJvm");
```

File: integrationTests/src/test/java/org/sonarqube/gradle/GradleTest.java [hunk 13]
```diff
@@ -332,4 +330,43 @@ public void testKotlinJvmWithSubmoduleProject() throws Exception {
     assertThat(binaries).containsExactly(baseDir.resolve("submodule/build/classes/kotlin/main").toString());
     assertThat(sources).containsExactly(baseDir.resolve("submodule/src/main/kotlin").toString());
   }
+
+  @Test
+  public void testScanAllOnMultiModuleWithSubModulesProject() throws Exception {
```
Comment (dorian-burihabwa-sonarsource) [inline integrationTests/src/test/java/org/sonarqube/gradle/GradleTest.java:335]: Can the title of the test be more explicit in what is expected to happen?
```suggestion
  public void testScanAllOnMultiModuleWithSubModulesProjectCollectsTheExpectedSources() throws Exception {
```

File: integrationTests/src/test/java/org/sonarqube/gradle/GradleTest.java [hunk 14]
```diff
@@ -332,4 +330,43 @@ public void testKotlinJvmWithSubmoduleProject() throws Exception {
...
+        baseDir.resolve("settings.gradle.kts").toString(),
+        baseDir.resolve("module/build.gradle.kts").toString(),
+        baseDir.resolve("module/submodule/build.gradle.kts").toString(),
+        baseDir.resolve("module/submodule/submoduleScript.sh").toString(),
+        baseDir.resolve("gradlew").toString(),
+        baseDir.resolve("gradlew.bat").toString(),
+        // Will be fixed by SCANGRADLE-149
+        baseDir.resolve("skippedModule/skippedSubmodule/build.gradle.kts").toString(),
+        baseDir.resolve("gradle/wrapper/gradle-wrapper.properties").toString())
+      .doesNotContain(
```
Comment (dorian-burihabwa-sonarsource) [inline integrationTests/src/test/java/org/sonarqube/gradle/GradleTest.java:362]: Let's move the closing parenthesis closer to the next line to make it clearer that  `.doesNotContain` starts a brand new assertion
```suggestion
        baseDir.resolve("gradle/wrapper/gradle-wrapper.properties").toString()
      ).doesNotContain(
```

File: integrationTests/src/test/resources/multi-module-with-submodules/settings.gradle.kts [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+        maven(url="maven-repo")
+        gradlePluginPortal()
+        ivy("ivy-repo")
+    }
+}
+
+include("module")
+include("skippedModule")
+include("module:submodule")
+include("skippedModule:skippedSubmodule")
```
Comment (dorian-burihabwa-sonarsource) [inline integrationTests/src/test/resources/multi-module-with-submodules/settings.gradle.kts:20]: So we have confirmation that we need to add submodule to the top level project?

File: src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java (hunk 4 above)
Comment (dorian-burihabwa-sonarsource) [reply src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java:482]: I do not think this closed here

File: src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java [hunk 16]
```diff
@@ -140,15 +160,91 @@ private void computeSonarProperties(Project project, Map<String, Object> propert
...
+      .map(File::toPath)
+      .collect(Collectors.toSet());
+
+    SourceCollector visitor = new SourceCollector(allModulesExistingSources, skippedDirs, Set.of(), false);
+
+    try {
+      Files.walkFileTree(project.getProjectDir().toPath(), visitor);
+    } catch (IOException e) {
+      LOGGER.error(String.valueOf(e));
+
```
Comment (dorian-burihabwa-sonarsource) [inline src/main/java/org/sonarqube/gradle/SonarPropertyComputer.java:226]: Let's remove the empty lin here
```suggestion
```

File: src/test/groovy/org/sonarqube/gradle/FunctionalTests.groovy
Comment (dorian-burihabwa-sonarsource) [inline src/test/groovy/org/sonarqube/gradle/FunctionalTests.groovy:1]: The diff is big enough as it is, let's try and restore the format of lines 1 to 357 (before the new tests).
If we want to re-format the code, let's agree on a groovy convention and apply the changes in a series of commits that do not contain functional changes

File: src/test/projects/java-multi-nested-modules/settings.gradle.kts [hunk 17]
```diff
@@ -0,0 +1,3 @@
+rootProject.name = "java-multi-module-with-skipped-modules"
```
Comment (dorian-burihabwa-sonarsource) [inline src/test/projects/java-multi-nested-modules/settings.gradle.kts:1]: We probably want to update the project name here to stay consistent

**Review by dorian-burihabwa-sonarsource — CHANGES_REQUESTED** [review]
Thanks for the changes, we are almost there 👍🏿

File: integrationTests/src/test/resources/multi-module-with-submodules/settings.gradle.kts (hunk 15 above)
Comment (ADarko22) [reply integrationTests/src/test/resources/multi-module-with-submodules/settings.gradle.kts:20]: Good question! You can include the modules in any `settings.gradle`, but you need to be careful because a module will apply the settings only from its closest ancestor!
In this case is fine to have a single settings file at the root and thus include all the modules there

**Review by dorian-burihabwa-sonarsource — APPROVED** [review]
